simpleman19/eagleEvents
# Feature/pd make seating arrangement

[VersionOne User Story](https://www7.v1host.com/OklahomaChristianUnivMba133/story.mvc/Summary?oidToken=Story%3A1293)


Get a seating arrangement by calling `SeatingChartGA(event).get_seating_chart_tables()`

## Testing
1. Import [this database](https://github.com/simpleman19/eagleEvents/blob/feature/pd-make-seating-arrangement/eagleEvents/testdb.sql) through the DB browser for SQLite (drop all tables first)
It uses the data from the 3,000 guest list csv  (I hopped on Jacob's branch for it, but had to do some massaging since the functionality was still WIP).

2. `python manage.py run_ga`
It generates a seating chart for the first event it pulls from the database (the sql script has 1 event).

3. Grab a cup of tea
3,000 guests with a total of 10,000 preferences takes a hot second, but this is the worst that we're supposed to handle. It took 57 seconds on my machine.

4. Look at that verbose output (mostly just the top)
You can see some stats on the fitnesses of each generation. Right now the fitness is the count of guests sitting at a table with someone that they dislike. So, a generation with a fitness minimum of 0 means that we found an optimal solution.

5. Try running it again with a smaller set
Execute this sql against the database:
```sql
delete from guest where number > 200
```

6. `python manage.py test` to run unit tests against `seating_chart_ga.py`

## Usage
Not sure if I did the database interactions correctly, so there may be something wrong here:
https://github.com/simpleman19/eagleEvents/blob/ed736de44c1557cd98b48fa5dfe7e6c67b6d7ce3/eagleEvents/models/event.py#L33-L43

## Missing requirement: `LIKE`
The genetic algorithm only accounts for dislikes right now. That means [this acceptance test](https://www7.v1host.com/OklahomaChristianUnivMba133/Test.mvc/Summary?oidToken=Test%3A1348) fails:
```
Given a guest wants to sit with another guest
When the genetic algorithm returns the final seating arrangement
Then the guest should be assigned to the same table as the other guest (unless conflicting constraints disallow this)
```
This won't be difficult to add on with [how DEAP handles multi-objective optimization](http://deap.readthedocs.io/en/master/tutorials/basic/part2.html#evaluation), but I don't know if I'll get it done by Monday (I'll also have to make changes for [multi-objective statistics](http://deap.readthedocs.io/en/master/tutorials/basic/part3.html#multi-objective-statistics)).

I'll add the functionality for `LIKE` in another PR. If it doesn't get done by Monday, then I'll make a defect for it in VersionOne and can tackle it in sprint 2 or 3.

## Other Improvements
None of these are required to demo MVP. But, if we have time later then maybe we can tackle some of these. The first two would be pretty easy.
- Add an objective that tries to minimize the number of tables that only have 1 guest assigned, so nobody sits alone
- Cache/keep the best individual from all generations (might be easy with DEAP's [Hall of Fame](http://deap.readthedocs.io/en/master/api/tools.html#hall-of-fame))
- Speedup with data parallelism (so far my attempts at using `multiprogramming` and `concurrent.futures` only slowed down execution time, ab72106 and ac93a76)
